app/domain/authentication/handler/authentication_handler.rb

```diff
@@ -130,7 +130,14 @@ def validate_client_ip(client_ip_address, conjur_role)
       end
 
       def fetch_conjur_role(authenticator, identity)
```

Comment:
I think we should consider moving this into either a `Role` or a `User` repository. This allows us to decouple the process of finding the relevant user from the authentication process (which this class does).

We could do something like this (_warning pseudocode_) :
```ruby
module DB
  module Repository
    class RoleRepository
      def initialize(role: ::Role, annotation: ::Annotation)
        @role = role
        @annotation = annotation
      end

      def find(account:, identifier:, annotation: nil, type: 'user')
        role = find_by_id(account: account, identifier: identifier, type: type)
        if role.blank? && annotation.present?
          role = find_by_annotation(account: account, identifier: identifier, type: type, annotation: annotation)
        end
        role
      end

      # The methods below are public to simplify testing, but only `find` should be part of the primary interface.

      def find_by_id(account:, identifier:, type: 'user')
        @role[[account, type, identifier].join(':')]
      end

      def find_by_annotation(account:, identifier:, annotation:, type: 'user')
        resource_id = @annotation
          .where(name: annotation)                    # exact match on annotation name
          .ilike(value: identifier)                   # case insensitive match on annotation value
          .like(resource_id: "#{account}:#{type}:%")  # verify account and type
          .first
          .try(:resource_id)
        
        return nil unless resource_id

        @role[resource_id]
      end
    end
  end
end
```

Alternatively, we can extend `Annotation` to include a method to handle finding a resource given an annotation name and value.  In this case, the `Annotation` method should return the `resource_id` to align with the `Role#roleid_from_username` convention.

Reply:
do you want a role repo or would this be fine to just add to the role model?

Reply:
The problem is `Annotation`.  Adding `find_by_annotation` to Role feels a bit weird as it forces `Annotation` as a dependency on `Role`. 

It might make more sense to add a `find_by_key_and_value` type method to `Annotation`.

Reply:
in that case should it be more like the original, but just have role_id be the default